numba/core/types/function_type.py

```diff
@@ -0,0 +1,189 @@
+                return types.unliteral(new_sig)
+
+        return sig
+
+    def check_signature(self, other_sig):
+        """Return True if signatures match (up to being precise).
+        """
+        sig = self.signature
+        return (self.nargs == len(other_sig.args)
+                and (sig == other_sig or not sig.is_precise()))
+
+    def unify(self, context, other):
+        if isinstance(other, types.UndefinedFunctionType) \
+           and self.nargs == other.nargs:
+            return self
+
+
+class UndefinedFunctionType(FunctionType):
+
+    _counter = 0
+
+    def __init__(self, nargs, dispatchers):
+        from numba.core.typing.templates import Signature
+        signature = Signature(types.undefined,
+                              (types.undefined,) * nargs, recvr=None)
+
+        super(UndefinedFunctionType, self).__init__(signature)
+
+        self.dispatchers = dispatchers
+
+        # make the undefined function type instance unique
+        type(self)._counter += 1
+        self._key += str(type(self)._counter)
+
+    def get_precise(self):
+        """
+        Return precise function type if possible.
+        """
+        for dispatcher in self.dispatchers:
+            for cres in dispatcher.overloads.values():
+                sig = types.unliteral(cres.signature)
+                return FunctionType(sig)
+        return self
+
+
+class FunctionPrototype(Type):
+    """
+    Represents the prototype of a first-class function type.
+    Used internally.
+    """
+    cconv = None
+
+    def __init__(self, rtype, atypes):
+        self.rtype = rtype
+        self.atypes = tuple(atypes)
+
+        assert isinstance(rtype, Type), (rtype)
+        lst = []
+        for atype in self.atypes:
+            assert isinstance(atype, Type), (atype)
+            lst.append(atype.name)
+        name = '%s(%s)' % (rtype, ', '.join(lst))
+
+        super(FunctionPrototype, self).__init__(name)
+
+    @property
+    def key(self):
+        return self.name
+
+
+class WrapperAddressProtocol(ABC):
+    """Base class for Wrapper Address Protocol.
+
+    Objects that inherit from the WrapperAddressProtocol can be passed
+    as arguments to Numba jit compiled functions where it can be used
+    as first-class functions. As a minimum, the derived types must
+    implement two methods ``__wrapper_address__`` and ``signature``.
+    """
+
+    @abstractmethod
+    def __wrapper_address__(self):
+        """Return the address of a first-class function.
+
+        Returns
+        -------
+        addr : int
+        """
+
+    @abstractmethod
+    def signature(self):
+        """Return the signature of a first-class function.
+
+        Returns
+        -------
+        sig : Signature
+          The returned Signature instance represents the type of a
+          first-class function that the given WrapperAddressProtocol
+          instance represents.
+        """
+
+
+class CompileResultWAP(WrapperAddressProtocol):
```

Comment:
This class doesn't seem to be used at all.

Reply:
This is true. Although it can be useful for turning dispatcher instance compilation results into first-class functions with precise types.

Reply:
To keep this, we need to have a test that exercises it and have docstrings.

Reply:
Added a new test `test_compilation_results` and docstrings to `CompileResultWAP`.